cpp/src/parquet/arrow/schema.cc

```diff
@@ -461,41 +552,39 @@ static Status GetTimestampMetadata(const ::arrow::TimestampType& type,
   }
 
   *physical_type = ParquetType::INT64;
-  *logical_type = LogicalTypeFromArrowTimeUnit(unit);
+  PARQUET_CATCH_NOT_OK(*annotation =
+                           TimestampAnnotationFromArrowTimestamp(type, target_unit));
 
-  // The user is requesting that all timestamp columns are casted to a specific
-  // type. Only 2 TimeUnit are supported by arrow-parquet.
+  // The user is explicitly asking for timestamp data to be converted to the
+  // specified units (target_unit).
   if (coerce) {
-    switch (unit) {
+    switch (target_unit) {
       case ::arrow::TimeUnit::MILLI:
       case ::arrow::TimeUnit::MICRO:
-        break;
       case ::arrow::TimeUnit::NANO:
+        break;
       case ::arrow::TimeUnit::SECOND:
         return Status::NotImplemented(
-            "Can only coerce Arrow timestamps to milliseconds"
-            " or microseconds");
+            "Can only coerce Arrow timestamps to milliseconds, microseconds, or "
+            "nanoseconds");
     }
-
     return Status::OK();
   }
 
-  // Until ARROW-3729 is resolved, nanoseconds are explicitly converted to
-  // int64 microseconds when deprecated int96 is not requested.
-  if (type.unit() == ::arrow::TimeUnit::NANO)
-    *logical_type = LogicalType::TIMESTAMP_MICROS;
-  else if (type.unit() == ::arrow::TimeUnit::SECOND)
+  // The user implicitly wants timestamp data to retain its original time units,
+  // however the Arrow seconds time unit can not be represented (annotated) in Parquet.
+  if (type.unit() == ::arrow::TimeUnit::SECOND) {
     return Status::NotImplemented(
-        "Only MILLI, MICRO, and NANOS units supported for Arrow timestamps with "
-        "Parquet.");
+        "Only MILLI, MICRO, and NANO units supported for Arrow timestamps with Parquet.");
+  }
```

Comment:
This code is reached if your have an Arrow timestamp in seconds and don't give any instructions on how to convert/coerce it.

Currently (before this patch), if you don't ask for int96 and you don't supply an explicit coercion instruction, you fall to here, where the converter is willing to truncate your Arrow nanoseconds to Parquet microseconds, but is unwilling (for some reason) to expand your Arrow seconds to, say, Parquet milliseconds.  I don't know the history or reasoning for this latter behavior, so my bias for conserving existing behavior whenever in doubt kicked in, and I retained it.

But it would be both easy and safe (it seems to me) to default to converting seconds to milliseconds (rather than failing) and I've done that in commit c2ac4cc1154bb0582fd9647b267959005e5b1681.  If you don't approve of this behavior change, we can (more or less) just revert that commit.

Reply:
Sounds fine to me